doc/man1/openssl-s_server.pod.in

```diff
@@ -781,6 +782,14 @@ instead of BIO_write() to send the HTTP response requested by a client.
 This option is only valid when B<-ktls> along with B<-WWW> or B<-HTTP>
 are specified.
 
+=item B<-no_zerocopy_sendfile>
+
+By default, SSL_sendfile() will use the zerocopy TX mode, which gives a
+performance boost when used with kTLS device offload. Note that invalid TLS
+records might be transmitted if the file is modified while being sent, so this
+option can be used to disable zerocopy. This option is only valid when
+B<-sendfile> is specified.
```

Comment:
It seems to me that by having this on by default you are introducing a breaking change. Stuff that worked before may no longer work. That's not allowed in a minor release, so it might be better to have it off by default and explicitly ask for it to be on.

Reply:
OpenSSL already supports zerocopy TLS sendfile on FreeBSD, and it's always active (i.e. there isn't even an option to disable it). The disclaimer that the file shouldn't be modified during sendfile already applies to FreeBSD to the same extent.

Since OpenSSL is a cross-platform library:

- The behavior on different platforms should be as close as possible, i.e. if OpenSSL on FreeBSD uses zerocopy by default, OpenSSL on Linux should use it too.
- I would assume that applications using a cross-platform library do not rely on implicit platform-specific behaviors, but stick to the strictest limitations among all platforms. As FreeBSD may corrupt TLS records if the application modifies the file during sendfile, applications using OpenSSL shouldn't modify the file if they care about this corruption. That means, it's safe to turn on the similar feature for Linux.

So, it's a breaking change only for the applications that are already broken (ones that modify the file during sendfile and don't want the corruption to happen, but when they are compiled on FreeBSD, they already suffer from it).

Having it off by default on Linux and always on on FreeBSD means introducing platform-specific behaviors exposed to the users of a cross-platform library, who expect it to behave the same way on all platforms.

Reply:
> I would assume that applications using a cross-platform library do not rely on implicit platform-specific behaviors,

I don't think this is a safe assumption. Although OpenSSL is a cross-platform library, not all users of OpenSSL care about its cross-platform capabilities.

> As FreeBSD may corrupt TLS records if the application modifies the file during sendfile, applications using OpenSSL shouldn't modify the file if they care about this corruption. 

Users on Linux are unlikely to be aware of this if their application does not run on FreeBSD.

Reply:
> That's not allowed in a minor release

What about merging it to the next major release? Is that allowed?

Reply:
> What about merging it to the next major release? Is that allowed?

In principle yes - but in practice there is no branch for it as yet. master is being used for 3.1.

We could choose to take this to OTC to determine whether this is sufficiently breaking to justify requiring that this is "off" by default.